### ISSUE TITLE
Different node input type for merged graph for causalLM models

### ISSUE THREAD

**Issue by claeyzre** [huggingface/optimum#1417]
### System Info

```shell
Optimum: 1.13.1
Torch: 2.0.1+cu117
```

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

While running this command to convert a CausalLM model into a merged FP16 onnx graph
`optimum-cli export onnx --model hf-internal-testing/tiny-random-GPTJForCausalLM  --task causal-lm-with-past --fp16 --for-ort --device cuda tinytesttorash
`

I later on opened my graph on netron.app and found something weird:
![image](https://github.com/huggingface/optimum/assets/30050914/2d195dc6-ca65-4c77-be30-62116099dc74)
(I cut some outputs)
As you can see, some output nodes needed to make the next pass, the **present.x.key** are in FP32, whereas the **present.x.value** are in FP16.
Also the inputs of the graph for the next pass needs all those keys to be actually FP16 (both key and value).
Is it something we should expect ? 

### Expected behavior

All output should be in FP16 to match the next pass inference.
Also the output logits might be in FP16 as well ?

**Comment by fxmarty** [comment]
Thank you @claeyzre for the report. This issue should impact only GPTJ and codegen, and should be fixed with https://github.com/huggingface/transformers/pull/26836 (once released / merged).

**Comment by fxmarty** [comment]
@claeyzre Regarding the dtype of `logits`, it is a valid comment indeed. From my understanding, there is sometime a cast to fp32 at the end of the model to match the original implementation, see e.g.:

https://github.com/facebookresearch/llama/blob/6b8cff0e0c5a379c3828e89691f4a5c352202643/llama/model.py#L484C13-L484C13
https://github.com/mistralai/mistral-src/blob/aa075d3c6acb0e4d1b041dea6f11967d56e231df/mistral/model.py#L213

It is unclear to me why some authors tend to include this cast in the model itself. In my opinion, if a cast is needed e.g. for loss computation, it should be outside. We can always patch the model on our side for the ONNX export to yield fp16 logits, but unfortunately this will not be fixed in Transformers.

**Comment by fxmarty** [comment]
@claeyzre Sorry for the delay. It is likely we get away fine with it thanks to https://github.com/huggingface/optimum/blob/15b8d1eed4d83c5004d3b60f6b6f13744b358f01/optimum/onnxruntime/modeling_ort.py#L638 & https://github.com/huggingface/optimum/blob/15b8d1eed4d83c5004d3b60f6b6f13744b358f01/optimum/onnxruntime/modeling_ort.py#L776-L783

**Comment by fxmarty** [comment]
@claeyzre Feel free to open a PR if you'd like to patch the export to remove the fp16 -> fp32 cast for specific models!
